arch/risc-v/src/common/espressif/esp_qencoder.c

```diff
@@ -0,0 +1,593 @@
+
+static struct esp_lowerhalf_s *esp_pcnt2lower(int pcnt)
+{
+  switch (pcnt)
+    {
+#ifdef CONFIG_ESP_PCNT_U0_QE
+    case 0:
+      return &g_pcnt0lower;
+#endif
+#ifdef CONFIG_ESP_PCNT_U1_QE
+    case 1:
+      return &g_pcnt1lower;
+#endif
+#ifdef CONFIG_ESP_PCNT_U2_QE
+    case 2:
+      return &g_pcnt2lower;
+#endif
+#ifdef CONFIG_ESP_PCNT_U3_QE
+    case 3:
+      return &g_pcnt3lower;
+#endif
+    default:
+      return NULL;
+    }
+}
+
+/****************************************************************************
+ * Name: esp_interrupt
+ *
+ * Description:
+ *   Common timer interrupt handling.  NOTE: Only 16-bit timers require timer
+ *   interrupts.
+ *
+ ****************************************************************************/
+
+#if 0 /* FIXME: To be implemented */
+static int esp_interrupt(int irq, void *context, void *arg)
+{
+  struct esp_lowerhalf_s *priv = (struct esp_lowerhalf_s *)arg;
+  uint16_t regval;
+
+  DEBUGASSERT(priv != NULL);
+
+  return OK;
+}
+#endif
+
+/****************************************************************************
+ * Name: esp_setup
+ *
+ * Description:
+ *   This method is called when the driver is opened.  The lower half driver
+ *   should configure and initialize the device so that it is ready for use.
+ *   The initial position value should be zero. *
+ *
+ ****************************************************************************/
+
+static int esp_setup(struct qe_lowerhalf_s *lower)
+{
+  struct esp_lowerhalf_s *priv = (struct esp_lowerhalf_s *)lower;
+  irqstate_t flags;
+  uint32_t regval;
+
+  /* Protected access to the registers */
+
+  flags = spin_lock_irqsave(&priv->lock);
+
+  /* Enable the PCNT Clock and Reset the peripheral */
+
+  periph_module_enable(PERIPH_PCNT_MODULE);
+  periph_module_reset(PERIPH_PCNT_MODULE);
+
+  /* Disable all events */
+
+  pcnt_ll_disable_all_events(priv->dev, priv->config->pcntid);
+
+  /* Configure the limits range PCNT_CNT_L_LIM | PCNT_CNT_H_LIM */
+
+  pcnt_ll_set_high_limit_value(priv->dev, priv->config->pcntid, INT16_MAX);
+  pcnt_ll_set_low_limit_value(priv->dev, priv->config->pcntid, INT16_MIN);
+
+  /* Setup POS/NEG/LCTRL/HCTRL/FILTER modes */
+
+  pcnt_ll_set_glitch_filter_thres(priv->dev, priv->config->pcntid, priv->config->filter_thres);
+  pcnt_ll_enable_glitch_filter(priv->dev, priv->config->pcntid, true);
+  pcnt_ll_set_edge_action(priv->dev, priv->config->pcntid, 0, PCNT_CHANNEL_EDGE_ACTION_DECREASE, PCNT_CHANNEL_EDGE_ACTION_INCREASE);
+  pcnt_ll_set_level_action(priv->dev, priv->config->pcntid, 0, PCNT_CHANNEL_LEVEL_ACTION_KEEP, PCNT_CHANNEL_LEVEL_ACTION_INVERSE);
+
+  pcnt_ll_set_edge_action(priv->dev, priv->config->pcntid, 1, PCNT_CHANNEL_EDGE_ACTION_INCREASE, PCNT_CHANNEL_EDGE_ACTION_DECREASE);
+  pcnt_ll_set_level_action(priv->dev, priv->config->pcntid, 1, PCNT_CHANNEL_LEVEL_ACTION_KEEP, PCNT_CHANNEL_LEVEL_ACTION_INVERSE);
+
+  /* Configure GPIO pins as Input with Pull-Up enabled */
```

Comment:
GPIO config and matrix settings can go in the initialization function. Those should be configured only once during startup.

Reply:
done